stellar-contract-env-host/src/host.rs

```diff
@@ -1342,27 +1342,57 @@ impl CheckedEnv for Host {
     }
 
     fn bigint_add(&self, x: Object, y: Object) -> Result<Object, HostError> {
-        let res = self.visit_obj(x, |a: &BigInt| self.visit_obj(y, |b: &BigInt| Ok(a.add(b))))?;
+        let res = self.visit_obj(x, |a: &BigInt| {
+            self.visit_obj(y, |b: &BigInt| {
+                a.checked_add(b)
```

Comment:
I know all they do is wrapping result in an `Option`. I think it's more favorable to call the checked version, if one is offered. The [documentation](https://docs.rs/num-bigint/latest/num_bigint/struct.BigUint.html#impl-CheckedAdd) mentions the possibility of overflow, in which case `None` will be returned. This just gives us an extra layer of safety without any added perf cost. If for whatever reason the operation does fail, we will trap the VM properly (with an error code) instead of just crashing.

Reply:
But they don't _actually_ do any overflow checking. One might argue that the overflow checking could be added in a future version. And this is true, but in that case it would be a protocol change anyway and we would need to deploy that carefully.

Reply:
I just think if they provided one (and specifically mentioned overflow being a possibility), we should use it in case some lower level dependency panics in ways we didn't expect it. This gives us more control over error emission. 
I un-did all the `checked_` ops. Since there isn't one for every function we call `num-bigint` (e.g. pow), so we are not fully protected either way. And it does make the code a bit cleaner.

Reply:
> I just think if they provided one (and specifically mentioned overflow being a possibility), we should use it in case some lower level dependency panics in ways we didn't expect it.

Let's turn to the code and see if this makes sense. Consider `checked_add` which has the implementation
```rust
    #[inline]
    pub fn checked_add(&self, v: &BigInt) -> Option<BigInt> {
        Some(self + v)
    }
```
What would happen if `self + v` were to panic? This function would panic, of course. The `Some` doesn't protect you from anything at all. `Some` _isn't_ error handling in its own right, you need to explicitly provide the error handling. Because `self + v` doesn't return a `Result` or some other type with error handling capabilities, the only thing it can do is panic on error. You mostly cannot catch panics in rust (see https://doc.rust-lang.org/std/panic/fn.catch_unwind.html for the exception, but it's not reliable). But even if you could, this code isn't trying to catch the panics either.

In summary, the returned `Option` is given you a false sense of security. And, again, if a later version added some error checking then that would be a protocol change anyway; we would not want to get that error checking automatically.

Am I making sense here?

Reply:
Yeah that makes sense. In that case, the right thing to do is to find and check for those conditions ourselves and call the "unchecked" version directly. Sorry I got confused about the panic part.